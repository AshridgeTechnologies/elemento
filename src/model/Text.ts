import Element, {ElementType} from './Element.js'

export default class Text implements Element {
    constructor(
        public id: string,
        public name: string,
        public contentExpr: string) {}

    static is(element: Element): element is Text {
        return element.constructor.name === this.name
    }

    kind: ElementType = 'Text'

}
