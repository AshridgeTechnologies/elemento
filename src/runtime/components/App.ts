import React, {createElement} from 'react'
import * as Elemento from '../index'

type Properties = {id: string, pages: { [key: string]: React.FunctionComponent<any>; }}

export default function App({id, pages}: Properties) {
    const defaultPage = Object.keys(pages)[0]
    const state = Elemento.useObjectStateWithDefaults(`${id}._data`, {currentPage: defaultPage})
    const {currentPage} = state
    const pagePath = id + '.' + currentPage

    return createElement('div', {id},
            createElement(pages[currentPage], {path: pagePath}))
}