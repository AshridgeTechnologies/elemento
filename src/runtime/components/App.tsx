import React, {createElement, useEffect} from 'react'
import {useGetObjectState} from '../appData'
import {Box, Button, Container, Typography} from '@mui/material'
import {SnackbarProvider, enqueueSnackbar, closeSnackbar} from 'notistack'
import {AppData} from './AppData'
import {noop} from '../../util/helpers'
import {isSignedIn, useSignedInState} from './authentication'
import {subscribeToNotifications, type Notification} from './notifications'

import {dndWrappedComponent} from './ComponentHelpers'

type Properties = {path: string, maxWidth?: string | number, fonts?: string[], startupAction?: () => void, children?: any, topChildren?: any}

const containerBoxCss = {
    height: '100%',
    paddingLeft: 2,
    paddingRight: 2,
    paddingTop: 1,
    paddingBottom: 1,
    marginTop: 0,
    marginBottom: 0,
}

const notify = (notification: Notification) => {
    const {level, message, details} = notification
    console.log('Notification', level, message, details)
    const snackbarMessage = <div>
        <Typography color='white'>{message}</Typography>
        <Typography color='white' variant='body2'>{details}</Typography>
    </div>
    enqueueSnackbar(snackbarMessage, {variant: level, persist: level === 'error'})
}

const insertFontLink = (fonts: string[]) => {
    if (fonts.length && !document.head.querySelector('link#web-font-link')) {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        const fontFamilies = fonts.map(font => `family=${font}`).join('&')
        link.href=`https://fonts.googleapis.com/css2?${fontFamilies}&display=swap`
        document.head.appendChild(link)
    }
}

const App: any = dndWrappedComponent(function App({path, maxWidth, fonts = [], startupAction = noop, children, topChildren}: Properties) {
    const state = useGetObjectState<AppData>(path)
    const {currentPage} = state
    const pagePath = path + '.' + currentPage.name

    useEffect( () => { insertFontLink(fonts) }, [] ) // wrap startupAction to ensure no result is returned to useEffect
    useEffect( () => { startupAction() }, [] ) // wrap startupAction to ensure no result is returned to useEffect
    useEffect(() => subscribeToNotifications( notify ), [])
    useSignedInState()
    const pageToDisplay = state.pageToDisplay(isSignedIn())
    return <Box id={path} display='flex' flexDirection='column' height='100%' width='100%'  className='ElApp'>
        <SnackbarProvider action={(snackbarId) => (
            <Button sx={{color: 'white'}} onClick={() => closeSnackbar(snackbarId)}>
                Dismiss
            </Button>
        )}/>
        <Box flex='0'>
            {topChildren}
        </Box>
        <Box flex='1' minHeight={0}>
            <Container maxWidth={false} sx={{...containerBoxCss, maxWidth}}>
                {createElement(pageToDisplay, {path: pagePath} as React.Attributes) /*we do not know the properties of each page*/}
            </Container>
        </Box>
        <Box flex='0'>
            {children}
        </Box>
    </Box>

})

App.State = AppData

export default App
