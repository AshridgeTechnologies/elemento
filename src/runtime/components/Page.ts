import React from 'react'
import {valueOfProps} from '../runtimeFunctions'
import {Stack} from '@mui/material'

type Properties = { path: string, children?: any }

export default function Page({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    return React.createElement(Stack, {
        id: path,
        justifyContent: 'flex-start',
        alignItems: 'flex-start',
        spacing: 2,
        height: '100%', width: '100%',
        sx: {overflowY: 'scroll', padding: 1, position: 'relative'},
        ...propVals,
        children
    })
}
