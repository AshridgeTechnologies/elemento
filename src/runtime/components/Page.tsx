import React from 'react'
import {valueOfProps} from '../runtimeFunctions'
import {Stack} from '@mui/material'

type Properties = {path: string, children?: any}

export default function Page({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    return <Stack id={path}
                  justifyContent="flex-start"
                  alignItems="flex-start"
                  spacing={2}
                  {...propVals}>
        {children}
    </Stack>
}
