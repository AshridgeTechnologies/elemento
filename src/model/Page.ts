import Element, {ElementType} from './Element'
import BaseElement,{equalArrays} from './BaseElement'
import Text from './Text'

export default class Page extends BaseElement implements Element {

    constructor(
        id: string,
        name: string,
        private readonly props: {style?: string},
        public readonly elements: ReadonlyArray<Element>
    ) {
        super(id, name)
    }

    kind: ElementType = 'Page'

    static is(element: Element): element is Page {
        return element.constructor.name === this.name
    }

    protected getElements(): ReadonlyArray<Element> { return this.elements }

    get properties() {return this.props}
    get style() { return this.props.style }

    set(id: string, propertyName: string, value: any): Page {
        if (id === this.id) {
            if (propertyName === 'name') {
                return new Page(this.id, value, this.props, this.elements)
            }
            if (propertyName === 'elements') {
                return new Page(this.id, this.name, this.props, value)
            }

            const updatedProps = {...this.props, [propertyName]:value}
            return new Page(this.id, this.name, updatedProps, this.elements)
        }

        const newElements = this.getElements().map( el => el.set(id, propertyName, value))
        if (!equalArrays(newElements, this.getElements())) {
            return new Page(this.id, this.name, this.props, newElements)
        }

        return this
    }

    static fromJSON({id, name, props, elements}: { id: string; name: string; props: any, elements: any[] }): Page {
        return new Page(id, name, props, elements.map(el => Text.fromJSON(el)))
    }
}