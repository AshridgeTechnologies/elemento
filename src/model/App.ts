import Page from './Page';

export default class App  {
    constructor(
        public id: string,
        public name: string,
        public pages: Page[]
    ) {}

}