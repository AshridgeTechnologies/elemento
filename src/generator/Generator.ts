import App from '../model/App.js'
import Element from '../model/Element.js'
import Page from '../model/Page.js'
import Text from '../model/Text.js'

export default class Generator {
    constructor(public app: App) {
    }

    outputFiles() {
        const page = this.app.pages[0];
        return [
            {
                name: 'appMain.js',
                content: this.appMainContent(page)
            }
        ]
    }

    private appMainContent(page: Page) {
        const children = page.elements.map(p => `        ${Generator.generateElement(p)},`).join('\n');
        return `function AppMain(props) {
    return React.createElement('div', null,
${children}
    )
}
`

    }

    private static generateElement(element: Element): string {
        if(Text.is(element)) return `React.createElement(TextElement, null, ${element.contentExpr})`
        return `'???'`
    }
}