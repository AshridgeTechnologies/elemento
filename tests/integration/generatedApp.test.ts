import renderer from 'react-test-renderer'
import React, {FunctionComponent} from 'react'
import TextElement from '../../src/runtime/TextElement.js'
import Text from '../../src/model/Text.js'
import App from '../../src/model/App.js'
import Generator from '../../src/generator/Generator.js'
import Page from '../../src/model/Page.js'

test('generated AppMain has correct contentExpr', ()=> {

    const app = new App('t1', 'test1', [
        new Page('p1', 'Page 1', [
            new Text('id1', 't1', '"Hi there!"'),
            new Text('id1', 't2', '2 + 2'),
        ])])

    const gen = new Generator(app)
    expect(gen.outputFiles()[0].name).toBe('appMain.js')
    const appMainCode = gen.outputFiles()[0].content
    const appMainBody = appMainCode.replace(/^function AppMain.*/m, '').replace(/^}$/m, '').trim()

    const AppMain = new Function(appMainBody) as FunctionComponent

    global.React = React
    // @ts-ignore
    global.TextElement = TextElement
    const component = renderer.create(React.createElement(AppMain))
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})