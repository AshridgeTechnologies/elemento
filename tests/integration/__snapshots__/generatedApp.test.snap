// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated AppMain has correct contentExpr 1`] = `
<div>
  <p>
    Hi there!
  </p>
  <p>
    4
  </p>
</div>
`;
